src/query/string_opfunc.c

```diff
@@ -3024,6 +3032,366 @@ db_string_elt (DB_VALUE * result, DB_VALUE * arg[], int const num_args)
+	    db_json_add_element_to_array (new_doc, double_value.data.d);
+	    pr_clear_value (&double_value);
+	  }
+	  break;
+	case DB_TYPE_JSON:
+	  db_json_add_element_to_array (new_doc, arg[i]->data.json.document);
+	  break;
+	}
+    }
+
+  str = db_json_get_raw_json_body_from_document (new_doc);
+  db_make_json (result, str, new_doc, true);
+
+  return NO_ERROR;
+}
+
+int
+db_json_insert (DB_VALUE * result, DB_VALUE * arg[], int const num_args)
+{
+  int i, error_code = NO_ERROR;
+  JSON_DOC *new_doc;
+  char *str;
+
+  if (num_args < 3)
+    {
+      DB_MAKE_NULL (result);
+      return NO_ERROR;
+    }
+
+  switch (arg[0]->domain.general_info.type)
+    {
+    case DB_TYPE_CHAR:
+      new_doc = db_json_get_json_from_str (arg[0]->data.ch.medium.buf, error_code);
+
+      if (error_code != NO_ERROR)
+	{
+	  assert (new_doc == NULL);
+	  return error_code;
+	}
+      break;
+    case DB_TYPE_JSON:
+      new_doc = db_json_get_copy_of_doc (arg[0]->data.json.document);
+      break;
+    }
+
+  for (i = 1; i < num_args; i += 2)
+    {
+      switch (arg[i + 1]->domain.general_info.type)
+	{
+	case DB_TYPE_CHAR:
+	  db_json_insert_func (new_doc, arg[i]->data.ch.medium.buf, arg[i + 1]->data.ch.medium.buf, error_code);
+	  break;
+	case DB_TYPE_JSON:
+	  db_json_insert_func (new_doc, arg[i]->data.ch.medium.buf, arg[i + 1]->data.json.document, error_code);
+	  break;
+	}
+
+      if (error_code != NO_ERROR)
+	{
+	  if (error_code == ER_INVALID_JSON)
+	    {
+	      db_json_delete_doc (new_doc);
+	      return error_code;
+	    }
+	  /* TODO should we simply ignore invalid paths or throw error? */
+	}
+    }
+
+  str = db_json_get_raw_json_body_from_document (new_doc);
+  db_make_json (result, str, new_doc, true);
+
+  return NO_ERROR;
+}
+
+int
+db_json_remove (DB_VALUE * result, DB_VALUE * arg[], int const num_args)
+{
+  int i, error_code;
+  JSON_DOC *new_doc;
+  char *str;
+
+  if (num_args < 2)
+    {
+      DB_MAKE_NULL (result);
+      return NO_ERROR;
+    }
+
+  switch (arg[0]->domain.general_info.type)
+    {
+    case DB_TYPE_CHAR:
+      new_doc = db_json_get_json_from_str (arg[0]->data.ch.medium.buf, error_code);
+
+      if (error_code != NO_ERROR)
+	{
+	  assert (new_doc == NULL);
+	  return error_code;
+	}
+      break;
+    case DB_TYPE_JSON:
+      new_doc = db_json_get_copy_of_doc (arg[0]->data.json.document);
+      break;
+    }
+
+  for (i = 1; i < num_args; i++)
+    {
+      db_json_remove_func (new_doc, arg[i]->data.ch.medium.buf, error_code);
+    }
+
+  str = db_json_get_raw_json_body_from_document (new_doc);
+  db_make_json (result, str, new_doc, true);
+
+  return NO_ERROR;
+}
+
+/*
+ * db_json_merge ()
+ * this function merges two by two json
+ * so merge (j1, j2, j3, j4) = merge_two (j1, (merge (j2, merge (j3, j4))))
+ * result (out): the merge result
+ * arg (in): the arguments for the merge function
+ * num_args (in)
+ */
+
+int
+db_json_merge (DB_VALUE * result, DB_VALUE * arg[], int const num_args)
+{
+  int i;
```

Comment:
I think current algorithm creates too many documents. What about something like this:

```cpp
JSON_DOC *accumulator = db_json_allocate_doc ();
for (i = 0; i < num_args; i++)
  {
    switch (db_value_type (&arg[i])
    {
    case DB_TYPE_JSON:
      { err = db_json_merge (db_get_json_doc (&arg[i]), accumulator); break; }
    case DB_TYPE_STRING:
      { err = db_json_convert_string_and_call (db_get_string (&arg[i]), db_json_merge, accumulator); break; }
    default: // error or assert false
    }
  }
...
db_make_json (result, str, accumulator, true);
```

db_json_merge should handle all document merging.

Reply:
please explain, the algorithm allocates only one document

Reply:
changed